### PR TITLE
wayland: rebased and merge into master

### PR DESCRIPTION
**Pull request by james-lawrence**
pull all the changes from wayland into master since we're down to just a few missing pieces in wayland (mostly pieces also missing within the x11 backend) time to let it soak in master.

### CODE REVIEW

**Comment by Maan2003** [comment]
We also want CI for wayland backend, probably just need to copy
https://github.com/linebender/druid/blob/2d80ca04312746567219ebed2aacc0ee5d72b6c7/.github/workflows/ci.yml#L113-L153

**Comment by james-lawrence** [comment]
agreed, can do it as part of the next PR if you want. CI env is very different than my environment, which isn't an issue but don't want to think about it for this one. =)

**Comment by Maan2003** [comment]
I can do it tomorrow. wanted it because `cargo clippy --features wayland --no-default-features` generates 107 warnings :sweat_smile:

**Comment by Maan2003** [comment]
can you add the following commits:
Maan2003@33b31ba08429a2c6963372ae3c7a866ee396a471 - add CI for wayland
Maan2003@3fe57bb23da4126ad243e77e396e7f4bfeebdf40 - fixes bug in build.rs
Maan2003@8a5884c6eaabc94a5f428175ae41c8d8605bcad2 - disable ApplicationExt static assertion for wayland
Maan2003@3c9591c47a73e3b3c059b27d03d00ccb5cd9dbb2 - fixes most of clippy warnings

**Comment by Maan2003** [comment]
also squash 11d72d8f027dd6a2fcea23691a3a3267fcb88ae7 and c3d7fdbe592c8cde6695ff346e034c5b14457222

**Comment by james-lawrence** [comment]
@Maan2003 all green! =)

**Review by emersion — CHANGES_REQUESTED** [review]
With my wlroots and Wayland maintainer hat on, I'm NACKing this pull request. Usage of wlr-output-management protocol is not suitable for Wayland clients like druid. We _will_ break your client if you go down this path.

**Comment by james-lawrence** [comment]
I talked with @emersion the path hes referring to is strictly how we're accessing screen dimensions, there is no alternative path but its not critical and the code will degrade gracefully. basically only screens output will be impacted. not really worried about it.

**Comment by emersion** [comment]
Sorry, it seems like there is a misunderstanding. Please remove all wlr-output-management usage. Regular clients like druid must not use it.

> the code will degrade gracefully

No. In the future, the compositor may disconnect the client with a fatal error if it tries to bind to the wlr-output-management interface.

**Comment by james-lawrence** [comment]
@emersion degrade gracefully means 'check if manager object exists' if it doesn't move on. leave it alone. not your project and I already have plans to remove it. just not in this PR.

**Comment by emersion** [comment]
> degrade gracefully means 'check if manager object exists' if it doesn't move on

Then be aware that your client _will_ break, ie. will be disconnected with a fatal error. We will advertise the manager global and kill any client that tries to use it.

I will not leave it alone, because it's a _very bad_ thing that regular clients use our privileged protocols.

**Comment by james-lawrence** [comment]
@emersion we talked, your concerns were understood. using unstable protocols by default comes with that risk. I'd expect the same potential risks from any of the unstable protocols. this isn't news to anyone. same with the xdg_output, or layershell.

**Comment by emersion** [comment]
No, it has nothing to do about the protocols being "unstable". "unstable" protocols (in wayland-protocols parlance) just means that a separate protocol may supersede it at some point, but "unstable" protocols never see backwards-incompatible changes. See the wayland-protocols README.

In other words, even if you use an "unstable" protocol, your client will never be broken by a protocol update.

What I'm talking about here is very different. I'm saying that I'll make it so your clients won't even launch on purpose, to avoid having other clients be tempted to use privileged protocols they must not use.

> this isn't news to anyone

Please take time to understand how Wayland works before making your points.

> if it breaks its entirely not your problem.

Yes, it _is_ my problem, because it causes more work for me. I'm already spending a good chunk of time trying to convince you and explain that this pull request is doing the wrong thing.

**Comment by kennylevinsen** [comment]
We have talked it out on IRC, and we're all good. :)

In the wlroots team we're mainly worried about special protocols ending up used by normal clients when such use will be blocked in the future. @james-lawrence explained that this part is going away in the next PR, so while we find it odd that its included here, I guess that's just people having different git workflows.

**Comment by james-lawrence** [comment]
@Maan2003 this is ready to go, I pulled in changes from the next branch I was going to PR to resolve @emersion concerns.

makes the PR larger than I wanted and makes it no longer a pure rebase but ./shrug. at this point I think with there will be 1 final small PR around layershell reinitializing unless I encountered something when working on my applications.

I'll be opening up issues around the remaining work once the final PR is in for documentation/disccusion purposes for whomever takes up the torch next. ;)

**Comment by james-lawrence** [comment]
oh and of course I'll be avail if issues crop up.

File: druid-shell/src/backend/wayland/display.rs
Comment (Maan2003) [inline druid-shell/src/backend/wayland/display.rs:125]: can you explain this in more detail:
- where is this Send and Sync is needed? 
- how can we be sure that we are not using it across threads?

File: druid-shell/src/backend/wayland/display.rs
Comment (Maan2003) [inline druid-shell/src/backend/wayland/display.rs:73]: I don't understand the purpose of this trait, and `GlobalEventSubscription` .

why can't `Dispatcher` just `BTreeMap<u64, Arc<dyn Fn(.......)>>` (using a type alias for `Arc<dyn ....>` for making it look readable)

File: druid-shell/src/backend/wayland/outputs/output.rs
Comment (Maan2003) [inline druid-shell/src/backend/wayland/outputs/output.rs:11]: this trait is no longer useful? There is only impl and just a single call to `consume`. that too in this module

File: druid-shell/src/backend/wayland/surfaces/idle.rs
Comment (Maan2003) [inline druid-shell/src/backend/wayland/surfaces/idle.rs:14]: also print the token id here? `IdleToken` impls `Debug`

File: druid-shell/src/backend/wayland/surfaces/surface.rs
Comment (Maan2003) [inline druid-shell/src/backend/wayland/surfaces/surface.rs:569]: some `use`s? very long types :sweat_smile:

File: druid-shell/src/backend/wayland/window.rs
Comment (Maan2003) [inline druid-shell/src/backend/wayland/window.rs:72]: again some `use`s, huge types :grimacing:

File: druid-shell/src/backend/wayland/window.rs
Comment (Maan2003) [inline druid-shell/src/backend/wayland/window.rs:65]: lifetime is not needed?

**Review by Maan2003 — COMMENTED** [review]
Sorry for such a late review :/

**Comment by Maan2003** [comment]
Also, choice of using `Rc` or `Arc` feels arbitrary. please explain the reasoning

**Comment by james-lawrence** [comment]
@Maan2003 all good, its holiday season. =)

**Comment by james-lawrence** [comment]
@Maan2003 generally speaking I recall some comment about moving to arc be default within druid but i could be misremembering.

I'm fairly indifferent to the usage of either outside of what rust itself enforces or performance related issues.

most (all?) of the RC usage is from the original branch I started this work off of. it seems slapdash because why I defaulted to using arcs and original branch defaulted to using RC.

**Comment by Maan2003** [comment]
> I recall some comment about moving to arc be default within druid.

that is probably for druid only, we mostly use `Rc` in druid shell

**Comment by james-lawrence** [comment]
ah then it was a misunderstanding on my end. is the arc/rc stuff a blocker for this PR? would be easier for me to tackle after this is merged to limit the diverging from my private branch.

File: druid-shell/src/backend/wayland/pointers.rs
Comment (james-lawrence) [reply druid-shell/src/backend/wayland/pointers.rs:52]: interesting. also not sure of the correct behavior. can fix if something crops up.

File: druid-shell/src/backend/wayland/pointers.rs
Comment (james-lawrence) [reply druid-shell/src/backend/wayland/pointers.rs:280]: will make this change but there are some issues with this code in general w/ respect to mutliple windows.

File: druid-shell/src/backend/wayland/display.rs
Comment (james-lawrence) [reply druid-shell/src/backend/wayland/display.rs:125]: entirely unnecessary now. will remove.

File: druid-shell/src/backend/wayland/outputs/mod.rs
Comment (james-lawrence) [reply druid-shell/src/backend/wayland/outputs/mod.rs:153]: can switch it to using new if you want. primarily using it for instantiation.

File: druid-shell/src/backend/wayland/display.rs
Comment (james-lawrence) [reply druid-shell/src/backend/wayland/display.rs:73]: original intention was to allow you to not rely on the particular implemention when processing events. I can probably remove it and collapse everything.

**Comment by Maan2003** [comment]
> is the arc/rc stuff a blocker for this PR? 

We can fix it later :)

**Review by Maan2003 — APPROVED** [review]
LGTM

File: druid-shell/examples/empty_window.rs
Comment (Maan2003) [inline druid-shell/examples/empty_window.rs:1]: Do we need this example? It is very similar to shello example

File: druid-shell/src/backend/mod.rs
Comment (Maan2003) [inline druid-shell/src/backend/mod.rs:43]: also add openbsd
```suggestion
#[cfg(all(feature = "wayland", any(target_os = "linux", target_os = "openbsd")))]
mod wayland;
#[cfg(all(feature = "wayland", any(target_os = "linux", target_os = "openbsd")))]
pub use wayland::*;
#[cfg(all(feature = "wayland", any(target_os = "linux", target_os = "openbsd")))]
pub(crate) mod shared;
```

File: druid-shell/examples/empty_window.rs
Comment (james-lawrence) [reply druid-shell/examples/empty_window.rs:1]: nope not really. will purge it.

**Comment by james-lawrence** [comment]
@Maan2003 bump =) weekend is coming up have some bug fixes in my branch I'm planning on opening another PR for =)

**Comment by Maan2003** [comment]
Sorry, I forgot to merge this.

**Pull request merged**
